include/net/websocket.h

```diff
@@ -0,0 +1,136 @@
+ *
+ * @details The function will automatically add websocket header to the
+ * message.
+ *
+ * @param ctx Websocket context.
+ * @param payload Websocket data to send.
+ * @param payload_len Length of the data to be sent.
+ * @param opcode Operation code (text, binary, ping, pong, close)
+ * @param mask Mask the data, see RFC 6455 for details
+ * @param final Is this final message for this message send. If final == false,
+ * then the first message must have valid opcode and subsequent messages must
+ * have opcode WS_OPCODE_CONTINUE. If final == true and this is the only
+ * message, then opcode should have proper opcode (text or binary) set.
+ * @param user_send_data User specific data to this connection. This is passed
+ * as a parameter to sent cb after the packet has been sent.
+ *
+ * @return 0 if ok, <0 if error.
+ */
+int ws_send_msg(struct http_ctx *ctx, u8_t *payload, size_t payload_len,
+		enum ws_opcode opcode, bool mask, bool final,
+		void *user_send_data);
+
+/**
+ * @brief Send message to client.
+ *
+ * @details The function will automatically add websocket header to the
+ * message.
+ *
+ * @param ctx Websocket context.
+ * @param payload Websocket data to send.
+ * @param payload_len Length of the data to be sent.
+ * @param opcode Operation code (text, binary, ping ,pong ,close)
+ * @param final Is this final message for this data send
+ * @param user_send_data User specific data to this connection. This is passed
+ * as a parameter to sent cb after the packet has been sent.
+ *
+ * @return 0 if ok, <0 if error.
+ */
+static inline int ws_send_msg_to_client(struct http_ctx *ctx,
+					u8_t *payload,
+					size_t payload_len,
+					enum ws_opcode opcode,
+					bool final,
+					void *user_send_data)
+{
+	return ws_send_msg(ctx, payload, payload_len, opcode, false, final,
+			   user_send_data);
+}
+
+/**
+ * @brief Strip websocket header from the packet.
+ *
+ * @details The function will remove websocket header from the network packet.
+ *
+ * @param pkt Network packet to send
+ * @param masked The mask status of the message is returned.
+ * @param mask_value The mask value of the message is returned.
+ *
+ * @return Websocket flag value is returned.
+ */
+u32_t ws_strip_header(struct net_pkt *pkt, bool *masked,
+		      u32_t *mask_value);
+
+/**
+ * @brief Mask or unmask a websocket message if needed
+ *
+ * @details The function will either add or remove the masking from the data.
+ *
+ * @param pkt Network packet to process
+ * @param masking_value The mask value to use.
+ */
+void ws_mask_pkt(struct net_pkt *pkt, u32_t masking_value);
+
+/**
+ * @brief This is called by HTTP server after all the HTTP headers have been
+ * received.
+ *
+ * @details The function will check if this is a valid websocket connection.
+ *
+ * @param parser HTTP parser instance
+ *
+ * @return 0 if ok, 1 if there is no body, 2 if HTTP connection is to be
+ * upgraded to websocket one
+ */
+int ws_headers_complete(struct http_parser *parser);
```

Comment:
Not sure if the function name corresponds to its functionality: "The function will check if this is a valid websocket connection.". If the function indeed does that, it should be named accordingly, otherwise, if name is right, does it have to be part of public API? (Maybe this header file needs to be split in "public API" and "private API" parts, if there's no choice of having 2 header files?)

Reply:
Some of the functions could be marked as internal as they are very specific to the handshake phase. I will separate them.